### PR TITLE
Add support for VL53L1X sensor

### PR DESCRIPTION
**Pull request by ruimarinho**
Add support for [VL53L1X Time-of-Flight Distance Sensor Carrier with Voltage Regulator, 400cm Max](https://www.pololu.com/product/3415).

### CODE REVIEW

File: sonoff/support.ino [hunk 1]
```diff
@@ -875,6 +875,10 @@ void GetFeatures()
 #ifdef USE_TM1638
   feature_sns1 |= 0x80000000;  // xsns_28_tm1638.ino
 #endif
+#ifdef USE_VL53L1X
+  feature_sns1 |= 0x90000000;  // xsns_29_vl53l1x.ino
```
Comment (ruimarinho) [inline sonoff/support.ino]: @arendst is this just an arbitrary number?

File: sonoff/user_config.h [hunk 2]
```diff
@@ -267,6 +267,7 @@
   #define USE_BMP                                // Add I2C code for BMP085/BMP180/BMP280/BME280 sensor (+4k code)
 //    #define USE_BME680                           // Add additional support for BME680 sensor using Bosch BME680 library (+4k code)
   #define USE_SGP30                              // Add I2C code for SGP30 sensor (+1k1 code)
+  #define USE_VL53L1X                              // Add I2C code for VL53L1X sensor
```
Comment (ruimarinho) [inline sonoff/user_config.h]: @arendst how do I get the compiled size code difference?

File: sonoff/xsns_29_vl53l1x.ino [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+  if (!sensor.init()) {
+    AddLog_P(LOG_LEVEL_DEBUG, PSTR("Unable to begin sensor"));
+    snprintf_P(log_data, sizeof(log_data), S_LOG_I2C_FOUND_AT, "VL53L1X", "Sensor not ready");
+    return;
+  }
+
+  AddLog_P(LOG_LEVEL_DEBUG, PSTR("Detected Vl53l1"));
+
+  sensor.setDistanceMode(VL53L1X::Long);
```
Comment (ruimarinho) [inline sonoff/xsns_29_vl53l1x.ino]: @arendst could this setting come from an enum from web settings? Could you provide me with guidance here?

File: sonoff/user_config.h (hunk 2 above)
Comment (arendst) [reply sonoff/user_config.h]: Compile once without the define and once with the define and substract the compiled file sizes.

File: sonoff/xsns_29_vl53l1x.ino (hunk 3 above)
Comment (arendst) [reply sonoff/xsns_29_vl53l1x.ino]: No webpages for sensors. Takes up too much space. If parameters are needed use the sensor29 command as used on the MHZ19 sensor

File: sonoff/support.ino (hunk 1 above)
Comment (arendst) [reply sonoff/support.ino]: Should be a bit value. As all bits of feature_sns1 have been used continue with feature_sns2 using bit mask 0x00000001

**Comment by ruimarinho** [comment]
@arendst all comments have been addressed. I haven't added commands yet - perhaps in a later phase.

**Comment by ruimarinho** [comment]
@arendst rebased again.

**Comment by brainnex** [comment]
hello, will it be here any progress in the near future? I want to use a VL53L0X but don't know if I should wait or take a HC-SR04 Ultrasonic Sensor.

**Comment by andrethomas** [comment]
I think there will be a merge soon but not in this current form.

Do not waste time with the HC-SR04 - It is totally useless for anything other than just detecting if something is present - measurements are not accurate at all!

Besides that, the pack of 10pcs I ordered 4 of them are reporting 1CM irrespective what is in front of it ;)

**Comment by botts7** [comment]
HI All,

Has there been any movement on this? would love to be able to try this out!

**Comment by gemu2015** [comment]
i have a running VL53L0X driver in my Universal fork =>https://github.com/gemu2015/Sonoff-Tasmota

it works very nice in the documented range of the device. however if you have no object in range the output temporary jumps to strange arbitrary values. (this is a known bug of the chip)

**Comment by botts7** [comment]
> i have a running VL53L0X driver in my Universal fork =>https://github.com/gemu2015/Sonoff-Tasmota
> 
> it works very nice in the documented range of the device. however if you have no object in range the output temporary jumps to strange arbitrary values. (this is a known bug of the chip)

Thank you! Ill give it a try!

**Comment by Martin-Dolan** [comment]
Looking forward to  VL53L0X being supported. 
So many cool uses.  
* Activate displays when nearby
* doorway counter / presence detection  (I am thinking two modules might be used in conjunction to see if people go IN or OUT of a room, then can be used for presence detection by knowing how many people in a room even without movement.)
* mailbox monitors

**Comment by gemu2015** [comment]
meanwhile the bug with the arbitrary values when no objects are in front of the sensor is solved. (it was an interference with a too small hole in the housing! make the hole much larger then the sensor itself!)

feel free to take it over to Tasmota from my fork =>https://github.com/gemu2015/Sonoff-Tasmota

**Comment by Nighthawk79** [comment]
Hello gemu2015,
would you explain how to use your code.
I have compiled your code and burned it on to Wemos d1. then i have choosen the I2c Pins, but i can't see anything from VL53L0X sensor.

Thanks in advance

**Comment by gemu2015** [comment]
@Nighthawk79 

did you enable the code with

#define USE_VL53L0X

in your user_config_overwrite.h ?

beside this and defining the i2c lines nothing else is needed

**Comment by Nighthawk79** [comment]
Hello gemu2015,

now i have did it and the wemos recognise the VL53L0X, but the sensor doesn't work at all.

it looks like this: VL53L0X Distance        0mm

**Comment by gemu2015** [comment]
what breakout board do you have, eventually you have to wire some more pins e.g. shutdown to vcc.

**Comment by Nighthawk79** [comment]
It's  GY-530, there is no shutdown pin, onli vcc, gnd, scl and sda.

**Comment by gemu2015** [comment]
ok then it should work out of the box. there is nothing to adapt in the driver.

sda,scl really ok?

vcc to 3,3V NOT to 5 V. this may kill the sensor

**Comment by Nighthawk79** [comment]
sda, scl ist ok and vcc is on 3,3
sensor is also ok, tested on Raspberry.

**Comment by gemu2015** [comment]
check in console with cmd i2cscan
turn log level to debug and show if sensor says => VL53L1X is ready

**Comment by gemu2015** [comment]
try updated version for better sensor check

**Comment by Nighthawk79** [comment]
hello gemu,

sorry, but it doesn't work for me.

here the Logoutput:

15:29:07 RSL: Received Topic /i2cscan, Data Size 0, Data 
15:29:07 RSL: Group 0, Index 1, Command I2CSCAN, Data 
15:29:07 RSL: stat/sonoff/RESULT = {"I2CScan":"Device(s) found at 0x29"}

**Comment by gemu2015** [comment]
try new version should work now

**Comment by gemu2015** [comment]
it was a callback that was changed in newer tasmota version. it now works again.

**Comment by Nighthawk79** [comment]
Unfortunately still doesn't work.

it recognize the sensor: 
23:10:50 CMD: i2cscan
23:10:50 RSL: stat/sonoff/RESULT = {"I2CScan":"Device(s) found at 0x29"}

but it shows only a init value:

VL53L0X Distance | 9999mm

**Comment by Martin-Dolan** [comment]
Have you taken the protective yellow cover off the sensor?

**Comment by Nighthawk79** [comment]
yes, i did.
As I said before, i have tested the sensor on a raspberry, sensor is ok.

**Comment by gemu2015** [comment]
the value 9999 means out of range. it is emitted on values of 0 or >2000 mm (supported range of sensor)

**Comment by Nighthawk79** [comment]
ok, now i have flashed my wemos one more time, and now it works, thank you for your great support.

it is possible to expand the range up to 3meters?

Greets 
Alex

**Comment by gemu2015** [comment]
the 53L0x has a maximum range of 2m while the 53L1x has a max range of 4 m

**Comment by Nighthawk79** [comment]
ah, ok.

thank you!

**Comment by Nighthawk79** [comment]
i have tested it now for some hours, the working-range is only ~1280mm, evry distance greater shows 9999mm

**Comment by gemu2015** [comment]
the 2 m range is only achievable with appropriate reflection surfaces. from the  data sheet only special white paper works with 2 m (table below)
Bildschirmfoto 2019-04-03 um 16 29 14

**Comment by Nighthawk79** [comment]
Thank you!

**Comment by FabianHM** [comment]
For me, the sensor works very well with the code

**Comment by FabianHM** [comment]
Was ist nun zu tun, damit der Sensor in die offizielle Tasmota Version aufgenommen wird??
Funktion ist derzeit einwandfrei und stabil.

**Comment by ascillato** [comment]
@FabianHM Please, can you translate that to English? Thanks

**Comment by joba-1** [comment]
he wants to know what is left to be done to get the code in

**Comment by hakan42** [comment]
"What is to be done to get this merged into the official release"

And, more importantly,

"The sensor works without issues and stable"

Actually, I would look forward to having this merged as well I plan on using this for salt solution level measurement :)

**Comment by SirRedZ** [comment]
How do I test this enhancement for the V53L1X sensor? Thanks for the pull request and efforts!

**Comment by andrethomas** [comment]
@SirRedZ clone from https://github.com/ruimarinho/Sonoff-Tasmota/tree/sensor/vl53l1x and test it that way - it is an older version of Tasmota though so don't try it on a critical device as you may need to flash it by hardwire if something fails during the downgrade/upgrade.

**Comment by SirRedZ** [comment]
> 
> 
> @SirRedZ clone from https://github.com/ruimarinho/Sonoff-Tasmota/tree/sensor/vl53l1x and test it that way - it is an older version of Tasmota though so don't try it on a critical device as you may need to flash it by hardwire if something fails during the downgrade/upgrade.

Thank you so much for this timely reply. I tried to compile myself and ended up with error after error, I'm sure it's my inexperience. I tried Arduino 1.8.7, GitPod + VSc/Platformio. No luck. If someone wanted to throw a pre-compiled (older sonoff I know) binary my way I would be greatful. Just interested in the V53L1X sensor for testing my module. Thanks! Great work!

[-=SrZ=-]

**Comment by meingraham** [comment]
Try https://gitpod.io#https://github.com/ruimarinho/Sonoff-Tasmota/tree/sensor/vl53l1x

**Comment by hakan42** [comment]
So, if I go to gitpod.io, I get a shell with the source cloned for me, right? How would I get a binary there?

**Comment by meingraham** [comment]
See https://github.com/arendst/Sonoff-Tasmota/wiki/Compiling-Tasmota-on-Gitpod

**Comment by hakan42** [comment]
> See https://github.com/arendst/Sonoff-Tasmota/wiki/Compiling-Tasmota-on-Gitpod

Yep, but I am getting and elf error on gitpod as well as on my laptop at home:

```
gitpod /workspace/Sonoff-Tasmota $ pio run

...

/home/gitpod/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld: .pio/build/sonoff/firmware.elf 

section `.text' will not fit in region `iram1_0_seg'
collect2: error: ld returned 1 exit status
*** [.pio/build/sonoff/firmware.elf] Error 1
```

Any idea what I am doing wrong here?

**Comment by meingraham** [comment]
Running out of RAM. You're going to have to tailor off some features. I link to a sample `user_config_override.h` in the Gitpod article. You can use that to decide what features you need and what you don't. For example, you don't need both Domoticz and Home Assistant. You may need neither. Also, you don't need all the sensors. Pick and choose. Then compile and see if that resolves the compilation error.

**Comment by andrethomas** [comment]
Here's a bin to test - I currently have no way to test it myself.

[sonoff.zip](https://github.com/arendst/Sonoff-Tasmota/files/3627867/sonoff.zip)

Compiled against the latest dev codebase using core 2.4.2 patched for PWM and Alexa.

**Comment by SirRedZ** [comment]
Thank you. Testing now!

**Comment by andrethomas** [comment]
> 00:00:00 I2C: VL53L1X found at 0x29.

Means the sensor was detected...

It should be visible on the WebUI and should report its measurement when teleperiod occurs so maybe set teleperiod to a lower value, say `teleperiod 60` then you should expect to see the JSON every 60 seconds.

**Comment by SirRedZ** [comment]
Closing as requested.

**Comment by andrethomas2** [comment]
Closing as suggested - Might move to experimental drivers on Playground at some point.

**Comment by DoTorr** [comment]
Got this to work on Tasmota 8.2.

Copied drivers from https://github.com/ruimarinho/Sonoff-Tasmota/blob/sensor/vl53l1x/

And updated references.  

Built tasmota-sensors on gitpod.

image

**Comment by DoTorr** [comment]
Here the files I had to modify on 8.2

image

**Comment by Tollbringer** [comment]
Thank you very much. I will look at this tonight and give it a go- very much appreciated.

Tollbringer

**Comment by apartmedia** [comment]
The Tasmota docs regarding the VL53L0X/VL53L1X has a note for adapting the laser operation of the sensor.
Is there any way to call the sensor api from within tasmota to set up or change the sensor to different modes? Does the sensor driver use some kind of fixed and compiled in standard configuration or is it changeable at runtime somehow?

**Comment by barbudor** [comment]
You read the note so you know how to do it
As the driver is written today, it's a static option at compile time
Easy to compile you own version using TasmoCompiler for example

**Comment by apartmedia** [comment]
> You read the note so you know how to do it
> As the driver is written today, it's a static option at compile time
> Easy to compile you own version using TasmoCompiler for example

Ok, thx for confirming…

**Pull request closed without merging**